### PR TITLE
Refactor RHCOS fetch API, support fetching qemu from partial config

### PR DESCRIPTION
**Pull request by cgwalters**
The *main* thing I'm really fixing here is that I have an installer
wrapper script: https://github.com/cgwalters/homegit/blob/master/bin/okdinst

And I noticed recently that I kept fetching old RHCOS versions; that's
of course because the image URL was hardcoded in my install config.  What
I really want is to fetch that dynamically.

So this PR changes the TF rendering code to do an on-demand fetch
if the image isn't filled out in the libvirt config.

Now further, as I was arguing in this PR:
https://github.com/openshift/installer/pull/710#issuecomment-440486784

The result of an install run from installer git master is currently a function
of the 3-tuple:
(installer, RHCOS, release payload)

I'd actually like to print out all 3 clearly in the logs; this is
a first step.

Today as far as I can see we do separate fetches in AWS for master,
bootstrap and worker, which is a bit broken.  Need to move that
to a higher level, and this will help with that.

### CODE REVIEW

File: pkg/types/libvirt/machinepool.go [hunk 1]
```diff
@@ -13,6 +13,8 @@ type MachinePool struct {
 	// Image is the URL to the OS image.
 	// E.g. "http://aos-ostree.rhev-ci-vms.eng.rdu2.redhat.com/rhcos/images/cloud/latest/rhcos-qemu.qcow2.gz"
 	Image string `json:"image"`
+	// Metadata string describing the image
+	ImageDescription string `json:"imageDescription,omitempty"`
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply pkg/types/libvirt/machinepool.go]: (Though libvirt is only for development/testing right?)

But anyways, I don't know the installer codebase well enough here - is there an easy way I can "thread through" extra metadata like this across the stages?

What if we added a InstallerPrivate field?

[the remark this answers is not in the document]
Comment (cgwalters) [reply pkg/types/libvirt/machinepool.go]: Or, another approach is to just move the "print OS version" log message to after each invocation of `rhcos.FetchBuild`.

**Comment by cgwalters** [comment]
OK split out https://github.com/openshift/installer/pull/866

However, some of the work in this PR to retain the structured metadata I think is still useful.

**Comment by cgwalters** [comment]
Though of course if we go to an [always-pinned](https://github.com/openshift/installer/pull/757) model this PR would be obsolete.

**Comment by cgwalters** [comment]
/hold cancel

OK, this is rebased :surfing_man: and while a lot of the code is the same, we landed the much more trivial https://github.com/openshift/installer/pull/866 in the meantime.

This PR now has a different primary purpose; see the updated commit message for details.

**Comment by cgwalters** [comment]
/hold

Will rebase this after https://github.com/openshift/installer/pull/902 lands.

**Comment by crawford** [comment]
@cgwalters after #902 lands, what does this PR get us?

File: pkg/tfvars/libvirt/cache.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ func (libvirt *Libvirt) UseCachedImage() (err error) {
 		return nil
 	}
 
-	logrus.Infof("Fetching OS image: %s", filepath.Base(libvirt.Image))
+	logrus.Infof("Fetching qemu URL: %s", filepath.Base(libvirt.Image))
```
Comment (wking) [inline pkg/tfvars/libvirt/cache.go]: nit: The log message should describe the rrsource being fetched, but "URL" is just *how* we're fetching it.  If you don't like "OS image", how about "QEMU image"?  Or "QCOW2" (although I don't know if we actually require this to be QCOW2 or if libvirt can handle other formats too)?

File: pkg/types/libvirt/machinepool.go (hunk 1 above)
Comment (wking) [inline pkg/types/libvirt/machinepool.go]: `MachinePool` is for user-supplied input (although we supply defaults when the user doesn't care).  This string doesn't seem like an actionable input.

**Comment by abhinavdahiya** [comment]
This looks abandoned. and installer has the images hardcoded for boot.

/close

**Pull request closed without merging**